### PR TITLE
proxy support

### PR DESCRIPTION
**Pull request by devildant**
hello, 

I added proxy management via environment variables if you're interested.

best regards

### CODE REVIEW

**Comment by ddsol** [comment]
Could you make the proxy option part of the regular options instead of using environment variables?

Also, for the module, this should be enough, but for the cli, there obviously needs to be a way to pass it in. Personally, I think it should be a command line option, with possibly environment variables as an alternate way to pass this in.

Finally, having a separate option for http and https shouldn't be necessary since we can detect which it is by parsing the proxy url.

Would you mind making these changes?

**Comment by devildant** [comment]
Hello, I will try to add the proxy via the options.
for management separate proxies http / https it's a standard, i've always seen these two environment variables, but i can add the protocol detection.

**Comment by devildant** [comment]
Hello,
i add the proxy options
i set priority : 
- proxy option
- HTTP_PROXY
- HTTPS_PROXY

i test and it's ok for me

**Comment by devildant** [comment]
Do you want another change?

File: index.js [hunk 1]
```diff
@@ -31,13 +31,85 @@ SOFTWARE.
...
+  , HttpsProxyAgent = require('https-proxy-agent')
   ;
 
 // These numbers were obtained by measuring and averaging both using this module and the official speedtest.net
 var speedTestDownloadCorrectionFactor = 1.135
   , speedTestUploadCorrectionFactor   = 1.139
   ;
 
+// 	Proxy by parameter
+var proxyOptions = null;
```
Comment (ddsol) [inline index.js]: Could you add these 3 variables to the var block above to follow consistent style?

File: index.js [hunk 2]
```diff
@@ -31,13 +31,85 @@ SOFTWARE.
...
+
+// Search a proxy (http|https) in env with case-insensitive
+var proxyHttpEnv = findPropertiesInEnvInsensitive("HTTP_PROXY");
+var proxyHttpsEnv = findPropertiesInEnvInsensitive("HTTPS_PROXY");
+
+function findPropertiesInEnvInsensitive(prop) {
+    prop = (prop + "").toLowerCase();
```
Comment (ddsol) [inline index.js]: The indentation style of this file should be 2 spaces.
There is no need for the `+ ""` because all call sites always pass a string.

File: index.js [hunk 3]
```diff
@@ -31,13 +31,85 @@ SOFTWARE.
...
+    for(var p in process.env){
```
Comment (ddsol) [inline index.js]: There should be a space between keywords (`for`, `if`) and `(` and between `)` and `{`:
```js
for (var p in process.env) {
  ...
  if (process.env.hasOwnProperty(p) && prop === p.toLowerCase()) {
    ...
  }
}
```

File: index.js [hunk 4]
```diff
@@ -31,13 +31,85 @@ SOFTWARE.
...
+        if(process.env.hasOwnProperty(p) && prop == (p+ "").toLowerCase()){
```
Comment (ddsol) [inline index.js]: `p` is already guaranteed to be a string, so it's not necessary to add the `+ ""`.

File: index.js [hunk 5]
```diff
@@ -31,13 +31,85 @@ SOFTWARE.
...
+}
+
+//set the proxy agent for each http request
+// priority : 
+// 1 - proxyOptions
+// 2 - proxyHttpEnv (HTTP_PROXY)
+// 3 - proxyHttpsEnv (HTTPS_PROXY)
+function proxy(options) {
+    if (proxyHttpEnv == null && proxyHttpsEnv == null && proxyOptions == null)
+    {
```
Comment (ddsol) [inline index.js]: Opening braces should be at the end of the line they are associated with.

File: index.js [hunk 6]
```diff
@@ -31,13 +31,85 @@ SOFTWARE.
...
+        return;
+    }
+    var proxy = null;
```
Comment (ddsol) [inline index.js]: All variables of a `function` should be defined at the start of a function in a single block.

File: index.js [hunk 7]
```diff
@@ -31,13 +31,85 @@ SOFTWARE.
...
+			proxy = proxyHttpsEnv;
+		}
+        // for http priority
+        if (proxyHttpEnv != null && !proxyHttpEnv.startsWith("https:")) {
+            haveHttp = true;
+        }
+    }
+
+    if (isSSL == false || haveHttp == true) {
+        var agent = new HttpProxyAgent(proxy);
```
Comment (ddsol) [inline index.js]: This variable should also be defined at the beginning of the function.
Alternatively, just assign the property directly:
```js
options.agent = new HttpProxyAgent(proxy);
```

File: package.json [hunk 8]
```diff
@@ -15,6 +15,8 @@
   "dependencies": {
     "chalk": "^1.1.3",
     "draftlog": "^1.0.12",
-    "xml2js": "^0.4.4"
+    "xml2js": "^0.4.4",
+	"http-proxy-agent": "^2.0.0",
```
Comment (ddsol) [inline package.json]: This should be properly indented.

File: index.js [hunk 9]
```diff
@@ -31,13 +31,85 @@ SOFTWARE.
...
+    {
+        if (proxyOptions.startsWith("https:")) {
+            isSSL = true
+        }
+        proxy = proxyOptions;
+    }
+    else {
+	// Test proxy by env
+        proxy = proxyHttpEnv;
+		if (proxyHttpEnv != null) {
```
Comment (ddsol) [inline index.js]: This should be properly indented with 2 spaces for each level.
There is no need to test for `null`, just use `if (proxyHttpEnv) {`.

File: bin/index.js [hunk 10]
```diff
@@ -170,7 +170,40 @@ updateLines();
 /*
  * Start speed test
  */
-var test = SpeedTestNet({maxTime: 10000});
+
```
Comment (ddsol) [inline bin/index.js]: Indentation in this file should also be 2 spaces per level.

File: bin/index.js [hunk 11]
```diff
@@ -170,7 +170,40 @@ updateLines();
...
+                errorParameter = "Error: bad parameters";
+            }
+        }
+    }
+});
+
+/*
+ * interrupt if error (parameters)
+ */
+if (errorParameter != null) {
```
Comment (ddsol) [inline bin/index.js]: There is no need to test for `null`, just use `if (errorParameter) {`

File: index.js [hunk 12]
```diff
@@ -31,13 +31,85 @@ SOFTWARE.
...
+    return null;
+}
+
+//set the proxy agent for each http request
+// priority : 
+// 1 - proxyOptions
+// 2 - proxyHttpEnv (HTTP_PROXY)
+// 3 - proxyHttpsEnv (HTTPS_PROXY)
+function proxy(options) {
+    if (proxyHttpEnv == null && proxyHttpsEnv == null && proxyOptions == null)
```
Comment (ddsol) [inline index.js]: There is no need to test for `null`, just use `if (!proxyHttpEnv && !proxyHttpsEnv && !proxyOptions) {`

File: index.js [hunk 13]
```diff
@@ -31,13 +31,85 @@ SOFTWARE.
...
+    {
+        return;
+    }
+    var proxy = null;
+    var isSSL = false;
+    var haveHttp = false;
+	// Test the proxy parameter first for priority
+    if (proxyOptions != null)
```
Comment (ddsol) [inline index.js]: There is no need to test for `null`, just use `if (proxyOptions) {`

File: index.js [hunk 14]
```diff
@@ -31,13 +31,85 @@ SOFTWARE.
...
+		if (proxyHttpEnv != null) {
+			//for support https in HTTP_PROXY env var
+			if (proxyHttpEnv.startsWith("https:"))
+			{
+				isSSL = true
+			} else {
+				haveHttp = true;
+			}
+
+		} else if (proxyHttpsEnv != null) {
```
Comment (ddsol) [inline index.js]: There is no need to test for `null`, just use `if (proxyHttpsEnv) {`.

File: index.js [hunk 15]
```diff
@@ -31,13 +31,85 @@ SOFTWARE.
...
+			isSSL = true
+			proxy = proxyHttpsEnv;
+		}
+        // for http priority
+        if (proxyHttpEnv != null && !proxyHttpEnv.startsWith("https:")) {
```
Comment (ddsol) [inline index.js]: Should this include the following line?
```js
  proxy = proxyHttpEnv;
```

Also, this should be properly indented with 2 spaces for each level.
There is no need to test for `null`, just use `if (proxyHttpEnv && !proxyHttpEnv.startsWith('https:')) {`.

File: index.js [hunk 16]
```diff
@@ -31,13 +31,85 @@ SOFTWARE.
...
+            haveHttp = true;
+        }
+    }
+
+    if (isSSL == false || haveHttp == true) {
```
Comment (ddsol) [inline index.js]: There is no need to test for `true` or `false`, just use `if (!isSSL || haveHttp) {`.

File: README.md [hunk 17]
```diff
@@ -61,6 +62,63 @@ The options include:
...
+$ set HTTP_PROXY=http://proxy:3128
+$ speedtest-net
+```
+
+Code use example for node:
+```js
+process.env["HTTP_PROXY"] =  process.env["http_proxy"] = "http://proxy:3128";
+//process.env["HTTPS_PROXY"] = process.env["https_proxy"] = "https://proxy:3128";
+
+const speedTest = require('speedtest-net');
```
Comment (ddsol) [inline README.md]: Since `speedtest` is quite old and uses `var` everywhere, I think it would be best not to use `const` in the docs. Also, using `var` makes this module compatible with older node versions (which may still run in many production environments).

Besides that, I'm all for `const` and `let`.

Comment (devildant) [reply README.md]: hello i respect your example : https://github.com/ddsol/speedtest.net

:)

Comment (ddsol) [reply README.md]: Yeah, my bad. If you want to, you can fix that one as well 😄.

File: package.json (hunk 8 above)
Comment (devildant) [reply package.json]: done

File: index.js (hunk 7 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 16 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 14 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 9 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 13 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 6 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 5 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 12 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 4 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 3 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 2 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 1 above)
Comment (devildant) [reply index.js]: done

File: bin/index.js (hunk 10 above)
Comment (devildant) [reply bin/index.js]: done

File: index.js (hunk 12 above)
Comment (devildant) [reply index.js]: done

File: README.md (hunk 17 above)
Comment (devildant) [reply README.md]: done

Comment (devildant) [reply README.md]: done

File: index.js (hunk 3 above)
Comment (devildant) [reply index.js]: oups copy/paste :p

Comment (devildant) [reply index.js]: done

File: index.js (hunk 15 above)
Comment (devildant) [reply index.js]: proxy = proxyHttpEnv; is declare by default at the beginning of the block, and when it is necessary I set the https with proxy = proxyHttpsEnv;

Comment (devildant) [reply index.js]: done

File: bin/index.js (hunk 11 above)
Comment (devildant) [reply bin/index.js]: done

**Comment by devildant** [comment]
normally it's ok now :)

File: index.js [hunk 18]
```diff
@@ -31,13 +31,78 @@ SOFTWARE.
...
   , speedTestUploadCorrectionFactor   = 1.139
+  , proxyOptions = null
   ;
 
+// Search a proxy (http|https) in env with case-insensitive
+var proxyHttpEnv = findPropertiesInEnvInsensitive("HTTP_PROXY");
+var proxyHttpsEnv = findPropertiesInEnvInsensitive("HTTPS_PROXY");
+
+function findPropertiesInEnvInsensitive(prop) {
+  prop = (prop).toLowerCase();
```
Comment (ddsol) [inline index.js]: No need for parentheses here

File: index.js [hunk 19]
```diff
@@ -31,13 +31,78 @@ SOFTWARE.
...
   ;
 
 // These numbers were obtained by measuring and averaging both using this module and the official speedtest.net
 var speedTestDownloadCorrectionFactor = 1.135
   , speedTestUploadCorrectionFactor   = 1.139
+  , proxyOptions = null
   ;
 
+// Search a proxy (http|https) in env with case-insensitive
+var proxyHttpEnv = findPropertiesInEnvInsensitive("HTTP_PROXY");
```
Comment (ddsol) [inline index.js]: Please add these variables to the block above

File: index.js [hunk 20]
```diff
@@ -31,13 +31,78 @@ SOFTWARE.
...
+  return null;
+}
+
+//set the proxy agent for each http request
+// priority : 
+// 1 - proxyOptions
+// 2 - proxyHttpEnv (HTTP_PROXY)
+// 3 - proxyHttpsEnv (HTTPS_PROXY)
+function proxy(options) {
+  var proxy = null;
```
Comment (ddsol) [inline index.js]: Please make this a single `var` statement

File: index.js [hunk 21]
```diff
@@ -31,13 +31,78 @@ SOFTWARE.
...
+  var isSSL = false;
+  var haveHttp = false;
+  
+  if (!proxyHttpEnv && !proxyHttpsEnv && !proxyOptions) {
+    return;
+  }
+  // Test the proxy parameter first for priority
+  if (proxyOptions)
+  {
+    if (proxyOptions.startsWith("https:")) {
```
Comment (ddsol) [inline index.js]: Would you mind dropping all `startsWith` calls? They are ES6.

File: index.js [hunk 22]
```diff
@@ -31,13 +31,78 @@ SOFTWARE.
...
+      isSSL = true
```
Comment (ddsol) [inline index.js]: Please make sure all lines are `;`-terminated

File: index.js [hunk 23]
```diff
@@ -31,13 +31,78 @@ SOFTWARE.
...
   ;
 
+// Search a proxy (http|https) in env with case-insensitive
+var proxyHttpEnv = findPropertiesInEnvInsensitive("HTTP_PROXY");
+var proxyHttpsEnv = findPropertiesInEnvInsensitive("HTTPS_PROXY");
+
+function findPropertiesInEnvInsensitive(prop) {
+  prop = (prop).toLowerCase();
+  for (var p in process.env) {
+    if (process.env.hasOwnProperty(p) && prop == (p).toLowerCase()) {
```
Comment (ddsol) [inline index.js]: No need for parentheses here `(p).toLowerCase()` => `p.toLowerCase()`

File: index.js (hunk 21 above)
Comment (devildant) [reply index.js]: by what do you want me to replace startwith? 

it is necessary to determine the protocol

File: index.js (hunk 20 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 18 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 19 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 22 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 23 above)
Comment (devildant) [reply index.js]: done

File: index.js (hunk 21 above)
Comment (ddsol) [reply index.js]: `if (proxyOptions.substr(0, 6) === "https:") {`

Comment (devildant) [reply index.js]: i will change that thx

Comment (devildant) [reply index.js]: done

**Comment by devildant** [comment]
updated :)
